cilium/CFP-27752-operator-manages-cilium-identities.md

```diff
@@ -0,0 +1,392 @@
+
+Operator managing CIDs is intended to work with `identity-allocation-mode=CRD`, which is the [default](https://github.com/cilium/cilium/blob/730cc49be6f500ba6b567fd782db3e58e0fa302e/install/kubernetes/cilium/values.yaml#L1508).
+
+All features that rely on using kvstore will not work until there is an equivalent kvstore implementation of Operator managing CIDs.
+
+
+
+* Cluster-mesh - KVStore is required for connecting multiple clusters, because the information about pod IPs and CIDs is shared with KVStore watchers.
+
+
+# Enablement
+
+The feature will be guarded by a flag `operator-manage-cid`, that can be set in container arguments in Operator’s and Agent’s manifests. The default values for both will be `false`, meaning that Operator doesn’t manage CIDs, while Agents do.
+
+
+## Release Operator first
+
+Releasing Operator first will allow us to first ensure that Operator code is compatible with Agents simultaneously managing CIDs. Then we can have higher confidence that upgrading to Agent’s not managing CIDs is safe.
+
+
+## Long term plan
+
+The long term plan is to make Operator managing CIDs the default, and deprecate the part of Agent that is currently used to manage CIDs.
+
+
+# Migration
+
+Any situation when neither Operator nor Agents are managing CIDs is effectively a downtime for network policy changes. The control plane side of network policies and other features relying on CIDs will not carry out any related changes during this time, causing them to stop working.
+
+
+## Upgrade
+
+Safe upgrade, without downtime, requires simultaneous CID management compatibility.
+
+If we upgrade the Operator from a version that is not managing CIDs to a version that is managing CIDs while Agents are also managing CIDs, there should be no negative impact on the system. Everything should work the same.
+
+This works well with the way KCP upgrade works. Operator is upgraded first when KCP nodes are restarted with the updated manifests. Agent daemonset updates with a rolling update, only after all KCP nodes are running and healthy with the new version of Operator. While Operator is already managing CIDs, Agent pods are restarted to a new version, and they stop managing CIDs. The new Agents will create CEPs with an empty (nil) assigned identity when there is no matching CID in watcher cache for a new pod. Operator will then take care of those CEPs, by creating and assigning new CIDs, on CEP watcher events and periodically on full reconciliation rounds.
+
+
+## Downgrade
+
+Downgrade without downtime is possible only if Agent daemonset is downgraded first -- updated to an earlier version where Agent manages CIDs. Otherwise, if KCP is downgraded first, there will be a downtime lasting at least the amount of time needed for Agent daemonset to be downgraded after the KCP downgrade.
```

Comment:
Worth noting that there are some aspects like this which I think do not entirely fit into the current upgrade docs under https://docs.cilium.io. Probably when we get to that point it'll be worth revisiting this point to understand exactly how we can best instruct users in a way that will minimize the pain for them.

That said I think this is also the sort of thing that we probably would want to have as part of maturity graduation criteria for the feature as it approaches stable.

Reply:
Sounds good. I would like to better understand the upgrade/downgrade criteria for Cilium. To find out what the guarantees and SLOs are.

It’s also worth expanding here on the meaning of downtime. Because here it only means that pods cannot start on nodes that are running cilium-agent that isn’t yet downgraded, and it doesn’t mean that traffic is disrupted in the cluster, or services don’t work (or even any other features).

On the first thought, I think @ovidiutirla's Dynamic Config(https://github.com/cilium/cilium/issues/34101) proposal shows great promise for such a case, to ensure safe downgrade/rollback/feature disablement.